src/context.rs

```diff
@@ -70,36 +70,61 @@ impl<'a> QueryMatchContext<'a> {
         }
     }
 
-    pub fn maybe_get_single_matching_node_for_query<'query, 'enclosing_node>(
+    pub fn maybe_get_single_captured_node_for_query<'query, 'enclosing_node>(
         &self,
         query: impl Into<ParsedOrUnparsedQuery<'query>>,
         enclosing_node: Node<'enclosing_node>,
+    ) -> Option<Node<'enclosing_node>> {
+        self.maybe_get_single_captured_node_for_filtered_query(query, |_| true, enclosing_node)
+    }
+
+    pub fn maybe_get_single_captured_node_for_filtered_query<'query, 'enclosing_node>(
+        &self,
+        query: impl Into<ParsedOrUnparsedQuery<'query>>,
+        mut predicate: impl FnMut(Node) -> bool,
+        enclosing_node: Node<'enclosing_node>,
     ) -> Option<Node<'enclosing_node>> {
         let query = query.into().into_parsed(self.language.language());
-        let mut query_cursor = QueryCursor::new();
-        let mut matches = query_cursor.matches(&query, enclosing_node, self.file_contents);
-        let first_match = matches.next()?;
-        if matches.next().is_some() {
-            return None;
-        }
-        let mut nodes_for_default_capture_index = first_match.nodes_for_capture_index(0);
-        let first_node = nodes_for_default_capture_index.next()?;
-        if nodes_for_default_capture_index.next().is_some() {
+        let captures = tree_sitter_grep::get_captures_for_enclosing_node(
+            self.file_contents,
+            &query,
+            0,
+            None,
+            enclosing_node,
+        );
+        let mut filtered_captures = captures
+            .filter_map(|capture_info| predicate(capture_info.node).then_some(capture_info.node));
+        let first_node = *filtered_captures.next()?;
```

Comment:
Sort of fascinating how the lifetime stuff is different for a "lending" iterator, eg here without this immediate dereference it yelled at me for the next line because basically the last item self-borrow would've still been "outstanding" when then trying to get the next one